### PR TITLE
[SPARK-13354] [SQL] push filter throughout outer join

### PR DESCRIPTION
**Pull request by davies**
For a query

```
select * from a left outer join b on a.a = b.a where b.b > 10
```

The condition `b.b > 10` will filter out all the row that the b part of it is empty.

In this case, we should use Inner join, and push down the filter into b.

### CODE REVIEW

**Comment by gatorsmile** [comment]
It sounds like this PR is related to the following two PRs https://github.com/apache/spark/pull/10567 and https://github.com/apache/spark/pull/10566

If we can convert the outer joins to inner joins, the push down will be done by the existing rules.

**Comment by davies** [comment]
@gatorsmile I will review those, you can take `canFilterOutNull` to #10567

**Comment by gatorsmile** [comment]
Thank you very much! @davies Will do it. : )

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -954,7 +980,7 @@ object PushPredicateThroughJoin extends Rule[LogicalPlan] with PredicateHelper {
         split(splitConjunctivePredicates(filterCondition), left, right)
 
       joinType match {
-        case Inner =>
+        case _ if isInnerJoin(joinType, leftFilterConditions, rightFilterConditions) =>
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I don't think we should be doing this kind of reasoning in this rule.  If you are going to convert outer joins to inner joins that should be its own rule.

**Comment by marmbrus** [comment]
If we are going to reason about null intolerance we should do that in a principled way (and there is already a start in the form of constraints) and not just mix it randomly into existing rules.

**Comment by davies** [comment]
@marmbrus I think #10567 make more sense, will pull out part of this into that one.

**Pull request closed without merging**